src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -3,49 +3,477 @@
+                            return;
+                        }
+                        case TF_DataType.TF_INT32:
+                        {
+                            var src = (int*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT32:
+                        {
+                            var src = (uint*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_INT64:
+                        {
+                            var src = (long*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT64:
+                        {
+                            var src = (ulong*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_DOUBLE:
+                        {
+                            var src = (double*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_FLOAT:
+                        {
+                            var src = (float*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+#endif
+                        default:
+                            throw new NotSupportedException();
+                    }
+                }
+            }
+        }
+
+        public static void CopyTo<T>(this Tensor tensor, Span<T> destination) where T : unmanaged
+        {
             unsafe
             {
-                var len = checked((int)tensor.size);
-                var src = (T*)tensor.buffer;
-                var span = new Span<T>(src, len);
-                span.CopyTo(values);
+                var len = checked((int) tensor.size);
+                //perform regular CopyTo using Span.CopyTo.
+                if (typeof(T).as_dtype() == tensor.dtype && tensor.dtype != TF_DataType.TF_STRING) //T can't be a string but tensor can.
+                {
+                    var src = (T*) tensor.buffer;
+                    var srcspan = new Span<T>(src, len);
```

Comment:
(nit) - variables should be camelCased. `srcspan` is not one word - it should be `srcSpan`.